contracts/contracts/lib/LibGateway.sol

```diff
@@ -548,4 +566,105 @@ library LibGateway {
+        if (applyType == IPCMsgType.TopDown || isLCA) {
+            (bool foundChildSubnetId, SubnetID memory childSubnetId) = toSubnetId.down(s.networkName);
+            if (!foundChildSubnetId) {
+                return CrossMessageValidationOutcome.InvalidDstSubnet;
+            }
+
+            (bool foundSubnet,) = LibGateway.getSubnet(childSubnetId);
+            if (!foundSubnet) {
+                return CrossMessageValidationOutcome.InvalidDstSubnet;
+            }
+        } else {
+            SubnetID memory commonParent = toSubnetId.commonParent(s.networkName);
+            if (commonParent.isEmpty()) {
+                return CrossMessageValidationOutcome.CommonParentNotExist;
+            }
+        }
+
+        return CrossMessageValidationOutcome.Valid;
+    }
+
+     // Function to map CrossMessageValidationOutcome to InvalidXnetMessageReason
+    function validationOutcomeToInvalidXnetMsgReason(CrossMessageValidationOutcome outcome) internal pure returns (InvalidXnetMessageReason) {
+        if (outcome == CrossMessageValidationOutcome.InvalidDstSubnet) {
+            return InvalidXnetMessageReason.DstSubnet;
+        } else if (outcome == CrossMessageValidationOutcome.CannotSendToItself) {
+            return InvalidXnetMessageReason.CannotSendToItself;
+        } else if (outcome == CrossMessageValidationOutcome.CommonParentNotExist) {
+            return InvalidXnetMessageReason.CommonParentNotExist;
+        }
+
+        revert("Unhandled validation outcome");
+    }
+    
+     /**
+     * @dev Propagates all the populated cross-net messages from the postbox.
+     */
+    function propagateAllPostboxMessages() internal {
+        GatewayActorStorage storage s = LibGatewayActorStorage.appStorage();
+
+        uint256 keysLength = s.postboxKeys.length();
+        bytes32[] memory ids = new bytes32[](keysLength);
+        for (uint256 i = 0; i < keysLength; ) {
+            bytes32 msgCid = s.postboxKeys.at(i);
+            ids[i] = msgCid;
+            LibGateway.propagatePostboxMessage(msgCid);
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+     /**
+     * @dev Propagates the populated cross-net message for the given `msgCid`.
+     * @param msgCid - the cid of the cross-net message
+     */
+    function propagatePostboxMessage(bytes32 msgCid) internal {
+        GatewayActorStorage storage s = LibGatewayActorStorage.appStorage();
+        IpcEnvelope storage crossMsg = s.postbox[msgCid];
+
+        if (crossMsg.isEmpty()) {
+            revert("Message not found in postbox");
+        }
+
+        bool shouldBurn = LibGateway.commitValidatedCrossMessage(crossMsg);
+
+        // Cache value before deletion to avoid re-entrancy
+        uint256 v = crossMsg.value;
+        bytes32 deterministicId = crossMsg.toDeterministicHash();
+
+        // Remove the message to prevent re-entrancy and clean up state
+        delete s.postbox[msgCid];
+        s.postboxKeys.remove(msgCid);
```

Comment:
this is buggy, if `msgCid` is removed from postboxKeys, the length of the set changes and this line: https://github.com/consensus-shipyard/ipc/pull/1157/files#diff-b26838f2cc5f8bf764313ca069618747d9171766dc768ece47de2d9c38ffddfbR632 will panic due to array out of bound.